pkg/trace/info/stats.go

```diff
@@ -155,15 +169,127 @@ func (ts *TagStats) publish() {
+	result.TraceIDZero = atomic.LoadInt64(&s.TraceIDZero)
+	result.SpanIDZero = atomic.LoadInt64(&s.SpanIDZero)
+	result.ForeignSpan = atomic.LoadInt64(&s.ForeignSpan)
+	return result
+}
+
+func (s *TracesDroppedStats) toMap() (result map[string]int64) {
+	return statsStructToMap(s.AtomicCopy(), "TracesDroppedStats")
+}
+
+func (s *TracesDroppedStats) String() string {
+	return inlineStatsMap(s.toMap())
+}
+
+// TracesMalformedStats contains counts for reasons malformed traces have been accepted after applying automatic fixes
+type TracesMalformedStats struct {
+	DuplicateSpanID       int64 `json:"duplicate_span_id,omitempty"`
+	ServiceEmpty          int64 `json:"service_empty,omitempty"`
+	ServiceTruncate       int64 `json:"service_truncate,omitempty"`
+	ServiceInvalid        int64 `json:"service_invalid,omitempty"`
+	SpanNameEmpty         int64 `json:"span_name_empty,omitempty"`
+	SpanNameTruncate      int64 `json:"span_name_truncate,omitempty"`
+	SpanNameInvalid       int64 `json:"span_name_invalid,omitempty"`
+	ResourceEmpty         int64 `json:"resource_empty,omitempty"`
+	TypeTruncate          int64 `json:"type_truncate,omitempty"`
+	InvalidStartDate      int64 `json:"invalid_start_date,omitempty"`
+	InvalidDuration       int64 `json:"invalid_duration,omitempty"`
+	InvalidHTTPStatusCode int64 `json:"invalid_http_status_code,omitempty"`
+}
+
+// AtomicCopy returns a safe copy of the struct by doing atomic reads from all fields
+func (s *TracesMalformedStats) AtomicCopy() (result TracesMalformedStats) {
+	result.DuplicateSpanID = atomic.LoadInt64(&s.DuplicateSpanID)
+	result.ServiceEmpty = atomic.LoadInt64(&s.ServiceEmpty)
+	result.ServiceTruncate = atomic.LoadInt64(&s.ServiceTruncate)
+	result.ServiceInvalid = atomic.LoadInt64(&s.ServiceInvalid)
+	result.SpanNameEmpty = atomic.LoadInt64(&s.SpanNameEmpty)
+	result.SpanNameTruncate = atomic.LoadInt64(&s.SpanNameTruncate)
+	result.SpanNameInvalid = atomic.LoadInt64(&s.SpanNameInvalid)
+	result.ResourceEmpty = atomic.LoadInt64(&s.ResourceEmpty)
+	result.TypeTruncate = atomic.LoadInt64(&s.TypeTruncate)
+	result.InvalidStartDate = atomic.LoadInt64(&s.InvalidStartDate)
+	result.InvalidDuration = atomic.LoadInt64(&s.InvalidDuration)
+	result.InvalidHTTPStatusCode = atomic.LoadInt64(&s.InvalidHTTPStatusCode)
+	return result
+}
+
+func (s *TracesMalformedStats) toMap() (result map[string]int64) {
+	return statsStructToMap(s.AtomicCopy(), "TracesDroppedStats")
+}
+
+func (s *TracesMalformedStats) String() string {
+	return inlineStatsMap(s.toMap())
 }
 
 // Stats holds the metrics that will be reported every 10s by the agent.
 // Its fields require to be accessed in an atomic way.
 type Stats struct {
 	// TracesReceived is the total number of traces received, including the dropped ones.
 	TracesReceived int64
-	// TracesDropped is the number of traces dropped.
-	TracesDropped int64
+	// TracesDropped contains stats about the count of dropped traces by reason
+	TracesDropped TracesDroppedStats
```

Comment:
What I meant here was to use [type embedding](https://golang.org/doc/effective_go.html#embedding), like this:
```suggestion
	TracesDropped
```
This means calling the type `TracesDropped` (without the suffix), and as a result of embedding, the field as well.

Reply:
Oooh I see. That's cool, had no clue this was possible. 

I just tried it out and I think it could hurt the readability of the code, because when incrementing some of these stats then you might not know whether the reason is related to dropped or malformed traces without jumping to where it's defined. 

Without the embedding it's clear:
* `atomic.AddInt64(&ts.TracesDropped.SpanIDZero, 1)`
* `atomic.AddInt64(&ts.TracesMalformed.ServiceEmpty, 1)`

With the embedding it gets a bit confusing:
* `atomic.AddInt64(&ts.SpanIDZero, 1)`
* `atomic.AddInt64(&ts.ServiceEmpty, 1)`

So maybe better to keep the encapsulation unless there's another reason to use the embedding?

Reply:
Oh but you don’t have to do the latter form. You can keep ts.TracesDropped.X even with embedding. I don’t think it works otherwise anyway, only for methods.

Reply:
It actually works like that, all the fields become embedded as well. If we're just going to use the fully qualified `ts.TracesDropped.X` then what's the advantage of using the embedding?

Reply:
I see. I honestly didn’t know that. I rarely use this feature. I just preferred the name without the suffix. You can go ahead and do whichever you prefer. Sorry to confuse

Reply:
Oh ok. I'll rename to remove `Stats` suffix, and leaving it to use encapsulation instead of embedding.